lib/at_cmd/at_cmd.c

```diff
@@ -171,124 +290,130 @@ static void socket_thread_fn(void *arg1, void *arg2, void *arg3)
 			goto next;
 		}
 
-		LOG_DBG("at_cmd_rx %d bytes, %s",
-			bytes_read, log_strdup(item->data));
+		LOG_DBG("at_cmd_rx %d bytes, %s", bytes_read, log_strdup(buf));
 
-		payload_len = get_return_code(item->data, &ret);
-
-		if (ret.state != AT_CMD_NOTIFICATION) {
-			if ((response_buf_len > 0) &&
-			    (response_buf != NULL)) {
-				if (response_buf_len >= payload_len) {
-					memcpy(response_buf, item->data,
-					       payload_len);
-				} else {
-					LOG_ERR("Response buffer not large "
-						"enough");
+		payload_len = get_return_code(buf, &ret);
 
-					ret.code  = -EMSGSIZE;
-				}
-
-				response_buf_len = 0;
-				response_buf     = NULL;
+		if (ret.state == AT_CMD_NOTIFICATION) {
+			ret.callback = notification_handler;
+		}
 
+		/* Allocate a buffer if we need to, verify its size otherwise */
+		if (ret.state == AT_CMD_NOTIFICATION ||
+		    (current_cmd.buf == NULL &&
+		     current_cmd.flags & AT_CMD_BUF_RESP)){
+			ret.buf = k_malloc(payload_len);
+			if (ret.buf == NULL) {
+				LOG_ERR("Failed to allocate response buffer");
+				ret.code  = -ENOMEM;
 				goto next;
 			}
+			ret.flags |= AT_RESP_ALLOC;
+		} else if (current_cmd.buf != NULL) {
+			if (current_cmd.buf_len < payload_len) {
+				LOG_ERR("Response buffer not large enough");
+				ret.code  = -EMSGSIZE;
+				goto next;
+			}
+			ret.buf = current_cmd.buf;
 		}
 
-		if (payload_len == 0) {
-			goto next;
-		}
-
-		if (ret.state == AT_CMD_NOTIFICATION) {
-			item->callback = notification_handler;
-		} else {
-			item->callback = current_cmd_handler;
+		/* Fill the buffer if there is one to fill at this point*/
+		if (ret.buf != NULL) {
+			memcpy(ret.buf, buf, payload_len);
 		}
 next:
-		/* If no callback was set, free the item.
-		 * Otherwise, work queue callback will free it.
-		 */
-		if (item->callback == NULL) {
-			k_mem_slab_free(&rsp_work_items, (void **)&item);
-		} else {
-			k_work_init(&item->work, callback_worker);
-			k_work_submit(&item->work);
-		}
-
-		/* Notify back only if command was sent. */
-		if ((k_sem_count_get(&cmd_pending) == 0) &&
-		    (ret.state != AT_CMD_NOTIFICATION)) {
-			current_cmd_handler = NULL;
+		distribute_resp(&current_cmd, &ret);
 
-			k_msgq_put(&return_code_msq, &ret, K_FOREVER);
+		/* We have now handled a command if it was not a notification */
+		if (ret.state != AT_CMD_NOTIFICATION) {
+			complete_cmd();
 		}
 	}
 }
 
-static inline int at_write(const char *const cmd, enum at_cmd_state *state)
+/* Allocate and copy cmd to a new buffer */
+static char *buffer_cmd(const char *cmd)
```

Comment:
Do you really need a function for this? You are calling it only once.

Reply:
Not at all. Integrated it into `at_cmd_write_with_callback`